Cadasta/cadasta-platform
# Async: Refactor download form to schedule task

### Proposed changes in this pull request

_NOTE: This PR works towards the goals of #1400_

The following changes were added:

1. We now rely on [cadasta-workertoolbox](https://github.com/Cadasta/cadasta-workertoolbox) to configure queues.
1. Added [drf-tmp-scoped-token](https://github.com/Cadasta/drf-tmp-scoped-token) as authentication mechanism for API.
1. Removed obsolete file download tooling, simplify the `DownloadForm`. Form now only controls valid export type choices, no download tooling needed.
1. Added signature for `export.export` task, add helper to schedule a project export.
1. Adjusted `ProjectDataDownload` view to handle `POST` requests by scheduling an export, messaging user with success, and redirecting user to ProjectDashboard view.

### How to test

If running server normally, scheduling the task will be a no-op. To some degree, we can trust that `<task>.apply_async(...)` will schedule a task, as this is standard [Celery functionality](http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.apply_async). However, if this is not acceptable, then you can run your dev server with `SQS=1` environment variable set. Assuming that your AWS credentials are configured in your dev machine (and that you have access to the SQS service), submitting the export will place a message in the queue.

The following bash helper function enables you to easily check the lengths of the SQS queues from the command line:

```sh
list-sqs-queues() {
  python -c "
import boto3
import json
print(
  json.dumps(
    [
      '{}:{}'.format(
        url.split('/')[-1],
        boto3.resource('sqs', 'us-west-2').Queue(url).attributes['ApproximateNumberOfMessages']
      ) for url in boto3.client('sqs', 'us-west-2').list_queues().get('QueueUrls', [])
    ], indent=2)
)"
}
```

### When should this PR be merged

This PR is a requirement for #1400.

### Risks

None foreseen. 

### Follow-up actions

See #1400 for followup tasks.

### Checklist (for reviewing)

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
- **Has the scalability of this change been evaluated?**
- **Is there a maintenance plan in place?**

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- **Are all UI and API inputs run through forms or serializers?** 
- **Are all external inputs validated and sanitized appropriately?**
- **Does all branching logic have a default case?**
- **Does this solution handle outliers and edge cases gracefully?**
- **Are all external communications secured and restricted to SSL?**

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.
